pkg/userdata/plugin/plugin.go

```diff
@@ -0,0 +1,105 @@
+//
+// Core UserData plugin.
+//
+
+// Package plugin provides the plugin side of the plugin mechanism.
+// Individual plugins have to implement the provider interface,
+// pass it to a new plugin instance, and call run.
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/kubermatic/machine-controller/pkg/apis/plugin"
+)
+
+// Provider defines the interface each plugin has to implement
+// for the retrieval of the userdata based on the given arguments.
+type Provider interface {
+	UserData(
+		spec clusterv1alpha1.MachineSpec,
+		kubeconfig *clientcmdapi.Config,
+		cloudConfig string,
+		cloudProviderName string,
+		clusterDNSIPs []net.IP,
+		externalCloudProvider bool,
+	) (string, error)
+}
+
+// Plugin implements a convenient helper to map the request to the given
+// provider and return the response.
+type Plugin struct {
+	provider Provider
+	debug    bool
+}
+
+// New creates a new plugin. Debug flag is not yet handled.
+func New(provider Provider, debug bool) *Plugin {
+	return &Plugin{
+		provider: provider,
+		debug:    debug,
+	}
+}
+
+// Run looks for the given request and executes it.
+func (p *Plugin) Run() error {
+	reqCmd := os.Getenv(plugin.EnvRequest)
+	switch reqCmd {
+	case plugin.EnvUserDataRequest:
```

Comment:
Can we maybe just use an `if req.CMD != EnvUserDataRequest { return someErr }\nreturn p.handleUserDataRequest`?

There is just this one case and its unlikely to become more, which is why the current approach of using a switch statement and factoring out the error handling in a distinct func seems needlessly complicated to me

Reply:
Issue also wanted a _Version_ request. Not yet sure about a use case for it. IMHO can add it later if needed. So I'll simplify it.